File: service-rest/src/main/java/com/bnorm/barkeep/rest/RestBookService.java
// Copyright 2016 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.rest;

import java.util.Collection;
import java.util.Set;

import javax.annotation.Nullable;

import org.jetbrains.annotations.NotNull;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RestController;

import com.bnorm.barkeep.db.DbBookService;
import com.bnorm.barkeep.db.DbUserService;
import com.bnorm.barkeep.model.Book;
import com.bnorm.barkeep.model.Recipe;
import com.bnorm.barkeep.model.User;
import com.bnorm.barkeep.service.BookService;
import com.bnorm.barkeep.service.UserService;
import com.fasterxml.jackson.annotation.JsonView;

@RestController
@RequestMapping("/books")
public class RestBookService extends AbstractRestService implements BookService {

  private final UserService userService;
  private final BookService bookService;

  @Autowired
  public RestBookService(DbUserService userService, DbBookService bookService) {
    this.userService = userService;
    this.bookService = bookService;
  }

  private Book findByOwner(long userId, long bookId) {
    Book book = bookService.getBook(bookId);
    if (!isOwnedBy(book, userId)) {
      throw new NotFound("Unable to find book with id=%d", bookId);
    }
    return book;
  }

  private Book findByUser(User user, long bookId) {
    Book book = bookService.getBook(bookId);
    Set<Book> books = user.getBooks();
    if (book == null || (books != null && !books.contains(book))) {
      throw new NotFound("Unable to find book with id=%d", bookId);
    }
    return book;
  }

  @JsonView(Book.class)
  @RequestMapping(method = RequestMethod.GET)
  @Override
  public Collection<Book> getBooks() {
    User user = userService.getUser(currentUser().getId());
    return user.getBooks();
  }

  @JsonView(Book.class)
  @RequestMapping(value = "/{bookId}", method = RequestMethod.GET)
  @Override
  public Book getBook(@PathVariable("bookId") long bookId) {
    User user = userService.getUser(currentUser().getId());
    return findByUser(user, bookId);
  }

  @JsonView(Recipe.class)
  @RequestMapping(value = "/{bookId}/recipes", method = RequestMethod.GET)
  public Set<Recipe> getBookRecipes(@PathVariable("bookId") long bookId) {
    Book book = getBook(bookId);
    return book.getRecipes();
  }

  @JsonView(Book.class)
  @RequestMapping(method = RequestMethod.POST)
  @Override
  public Book createBook(@RequestBody Book book) {
    if (book.getId() != null) {
      throw new BadRequest("Cannot create book with existing id=%d", book.getId());
    }
    User currentUser = currentUser();
    if (book.getOwner() == null) {
      return bookService.createBook(new BookWithOwner(book, currentUser));
    } else if (!isOwnedBy(book, currentUser.getId())) {
      throw new BadRequest("Cannot create book owned by another user");
    } else {
      return bookService.createBook(book);
    }
  }

  @JsonView(Book.class)
  @RequestMapping(value = "/{bookId}", method = RequestMethod.PUT)
  @Override
  public Book setBook(@PathVariable("bookId") long bookId, @RequestBody Book book) {
    findByOwner(currentUser().getId(), bookId);
    return bookService.setBook(bookId, book);
  }

  @RequestMapping(value = "/{bookId}", method = RequestMethod.DELETE)
  @Override
  public void deleteBook(@PathVariable("bookId") long bookId) {
    findByOwner(currentUser().getId(), bookId);
    bookService.deleteBook(bookId);
  }

  private static class BookWithOwner implements Book {

    private final Book book;
    private final User owner;

    private BookWithOwner(Book book, User owner) {
      this.book = book;
      this.owner = owner;
    }

    @Override
    public Long getId() {
      return book.getId();
    }

    @Nullable
    @Override
    public String getTitle() {
      return book.getTitle();
    }

    @Nullable
    @Override
    public String getDescription() {
      return book.getDescription();
    }

    @Override
    public User getOwner() {
      return owner;
    }

    @Override
    public Set<Recipe> getRecipes() {
      return book.getRecipes();
    }

    @Override
    public int compareTo(@NotNull Book other) {
      return Book.Companion.getCOMPARATOR().compare(this, other);
    }
  }
}

File: service-rest/src/main/kotlin/com/bnorm/barkeep/rest/RestBookService.kt
// Copyright 2017 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.rest

import com.bnorm.barkeep.db.DbBookService
import com.bnorm.barkeep.model.Book
import com.bnorm.barkeep.model.BookSpec
import com.bnorm.barkeep.model.Recipe
import com.bnorm.barkeep.model.User
import com.bnorm.barkeep.service.BookService
import com.fasterxml.jackson.annotation.JsonView
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/books")
class RestBookService(private val bookService: DbBookService,
                      private val userService: RestUserService,
                      private val recipeService: RestRecipeService) : BookService {

  private fun isOwnedBy(book: BookSpec, userId: Long): Boolean = book.owner?.id == userId

  private fun assertAccess(book: Book) {
    val user = userService.getUser()
    val books = user.books
    if (!books.contains(book)) {
      throw NotFound(MSG_NOT_FOUND(TYPE_BOOK, book.id))
    }
  }

  private fun assertOwnership(book: Book) {
    if (!isOwnedBy(book, userService.currentId())) {
      throw Forbidden(MSG_DO_NOT_OWN(TYPE_BOOK, book.id))
    }
  }

  @JsonView(Book::class)
  @GetMapping
  override fun getBooks(): Collection<Book> {
    val user = userService.getUser()
    return user.books
  }

  @JsonView(Book::class)
  @GetMapping("/{bookId}")
  override fun getBook(@PathVariable("bookId") id: Long): Book {
    val book = bookService.findBook(id) ?: throw NotFound(MSG_NOT_FOUND(TYPE_BOOK, id))
    assertAccess(book)
    return book
  }

  @JsonView(Recipe::class)
  @GetMapping("/{bookId}/recipes")
  fun getBookRecipes(@PathVariable("bookId") id: Long): Set<Recipe> {
    val book = getBook(id)
    return book.recipes
  }

  @JsonView(Book::class)
  @PostMapping
  override fun createBook(@RequestBody book: BookSpec): Book {
    val currentUser = userService.getUser()
    if (book.owner == null) {
      return bookService.createBook(BookWithOwner(book, currentUser))
    } else if (!isOwnedBy(book, currentUser.id)) {
      throw BadRequest(MSG_CREATE_WRONG_USER("book"))
    } else {
      return bookService.createBook(book)
    }
  }

  @PostMapping("/{bookId}/recipes/{recipeId}")
  override fun addBookRecipe(@PathVariable("bookId") bookId: Long, @PathVariable("recipeId") recipeId: Long) {
    val book = getBook(bookId)
    val recipe = recipeService.getRecipe(recipeId)
    bookService.addBookRecipe(book.id, recipe.id)
  }

  @DeleteMapping("{/bookId}/recipes/{recipeId}")
  override fun removeBookRecipe(@PathVariable("bookId") bookId: Long, @PathVariable("recipeId") recipeId: Long) {
    val book = getBook(bookId)
    val recipe = recipeService.getRecipe(recipeId)
    bookService.removeBookRecipe(book.id, recipe.id)
  }

  @JsonView(Book::class)
  @PutMapping("/{bookId}")
  override fun setBook(@PathVariable("bookId") id: Long, @RequestBody book: BookSpec): Book {
    assertOwnership(getBook(id))
    return bookService.setBook(id, book)
  }

  @DeleteMapping("/{bookId}")
  override fun deleteBook(@PathVariable("bookId") id: Long) {
    assertOwnership(getBook(id))
    bookService.deleteBook(id)
  }

  private class BookWithOwner(private val book: BookSpec, override val owner: User) : BookSpec {
    override val title: String?
      get() = book.title
    override val description: String?
      get() = book.description
  }
}
